velox/dwio/parquet/reader/DeltaByteArrayDecoder.h

```diff
@@ -0,0 +1,190 @@
+  std::unique_ptr<DeltaBpDecoder> lengthDecoder_;
+  int32_t numValidValues_{0};
+  uint32_t lengthIdx_{0};
+  std::vector<uint32_t> bufferedLength_;
+};
+
+// DeltaByteArrayDecoder is adapted from Apache Arrow:
+// https://github.com/apache/arrow/blob/apache-arrow-15.0.0/cpp/src/parquet/encoding.cc#L3301-L3545
+class DeltaByteArrayDecoder {
+ public:
+  explicit DeltaByteArrayDecoder(const char* start) {
+    prefixLenDecoder_ = std::make_unique<DeltaBpDecoder>(start);
+    int64_t numPrefix = prefixLenDecoder_->validValuesCount();
+    bufferedPrefixLength_.resize(numPrefix);
+    prefixLenDecoder_->readValues<uint32_t>(
+        bufferedPrefixLength_.data(), numPrefix);
+    prefixLenOffset_ = 0;
+    numValidValues_ = numPrefix;
+
+    suffixDecoder_ = std::make_unique<DeltaLengthByteArrayDecoder>(
+        prefixLenDecoder_->bufferStart());
+    lastValue_.clear();
+  }
+
+  void skip(uint64_t numValues) {
+    skip<false>(numValues, 0, nullptr);
+  }
+
+  template <bool hasNulls>
+  inline void skip(int32_t numValues, int32_t current, const uint64_t* nulls) {
+    if (hasNulls) {
+      numValues = bits::countNonNulls(nulls, current, current + numValues);
+    }
+    for (int32_t i = 0; i < numValues; ++i) {
+      readString();
+    }
+  }
+
+  template <bool hasNulls, typename Visitor>
+  void readWithVisitor(const uint64_t* nulls, Visitor visitor) {
+    int32_t current = visitor.start();
+    skip<hasNulls>(current, 0, nulls);
+    int32_t toSkip;
+    bool atEnd = false;
+    const bool allowNulls = hasNulls && visitor.allowNulls();
+    for (;;) {
+      if (hasNulls && allowNulls && bits::isBitNull(nulls, current)) {
+        toSkip = visitor.processNull(atEnd);
+      } else {
+        if (hasNulls && !allowNulls) {
+          toSkip = visitor.checkAndSkipNulls(nulls, current, atEnd);
+          if (!Visitor::dense) {
+            skip<false>(toSkip, current, nullptr);
+          }
+          if (atEnd) {
+            return;
+          }
+        }
+
+        // We are at a non-null value on a row to visit.
+        toSkip = visitor.process(readString(), atEnd);
+      }
+      ++current;
+      if (toSkip) {
+        skip<hasNulls>(toSkip, current, nulls);
+        current += toSkip;
+      }
+      if (atEnd) {
+        return;
+      }
+    }
+  }
+
+  folly::StringPiece readString() {
+    auto suffix = suffixDecoder_->readString();
+    bool isFirstRun = (prefixLenOffset_ == 0);
+    const int64_t prefixLength = bufferedPrefixLength_[prefixLenOffset_++];
+
+    VELOX_CHECK_GE(
+        prefixLength, 0, "negative prefix length in DELTA_BYTE_ARRAY");
+
+    folly::StringPiece prefix{lastValue_};
+
+    buildBuffer(isFirstRun, prefixLength, suffix, prefix);
+
+    numValidValues_--;
+    lastValue_ = std::string{prefix};
+    return {lastValue_};
+  }
+
+ private:
+  void buildBuffer(
+      bool isFirstRun,
+      const int64_t prefixLength,
+      folly::StringPiece suffix,
+      folly::StringPiece& prefix) {
```

Comment:
This parameter is also not needed, just initialize it inside the function `folly::StringPiece prefix = lastValues_`

Reply:
@Yuhta I just removed the `prefix` variable and directly used `lastValue_` for the operation.